EIPS/eip-draft-blake2b-f-precompile.md

```diff
@@ -0,0 +1,76 @@
+---
+eip: <to be assigned>
+title: Add Blake2 compression function `F` precompile
+author: Tjaden Hess <tah83@cornell.edu>, Matt Luongo (@mhluongo)
+discussions-to: https://ethereum-magicians.org/t/blake2b-f-precompile/3157
+status: Draft
+type: Standards Track
+category: Core
+created 2016-10-04
+requires: 2046
+---
+
+<!--You can leave these HTML comments in your merged EIP and delete the visible duplicate text guides, they will not appear and may be helpful to refer to if you edit it again. This is the suggested template for new EIPs. Note that an EIP number will be assigned by an editor. When opening a pull request to submit your EIP, please use an abbreviated title in the filename, `eip-draft_title_abbrev.md`. The title should be 44 characters or less.-->
+
+## Simple Summary
+<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the EIP.-->
+
+This EIP will enable the Blake2b hash function to run cheaply on the EVM, allowing easier interoperability between Ethereum and Zcash as well as other Equihash-based PoW coins.
+
+## Abstract
+<!--A short (~200 word) description of the technical issue being addressed.-->
+
+This EIP introduces a new precompiled contract which implements the compression function `F` used in the BLAKE2b cryptographic hashing algorithm, for the purpose of allowing interoperability between the EVM and Zcash, as well as introducing more flexible cryptographic hash primitives to the EVM.
+
+## Motivation
+<!--The motivation is critical for EIPs that want to change the Ethereum protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the EIP solves. EIP submissions without sufficient motivation may be rejected outright.-->
+
+Besides being a useful cryptographic hash function and SHA3 finalist, BLAKE2b allows for efficient verification of the Equihash PoW used in Zcash, making a BTC Relay - style SPV client possible on Ethereum. A single verification of an Equihash PoW verification requires 512 iterations of the hash function, making verification of Zcash block headers prohibitively expensive if a Solidity implementation of BLAKE2b is used.
+
+The BLAKE2b algorithm is highly optimized for 64-bit CPUs, and is faster than MD5 on modern processors.
+
+Interoperability with Zcash could enable contracts like trustless atomic swaps between the chains, which could provide a much needed aspect of privacy to the very public Ethereum blockchain.
+
+## Specification
```

Comment:
Discussing gas price based on [benchmarks from the implementation](https://github.com/keep-network/go-ethereum/pull/4#discussion_r295574236) and using the same approach as we did for [EIP-1108](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1108.md)...

Let's assume `ecrecover` is perfectly priced. It costs `3000` gas and takes `180.3` microseconds on my Mac to execute. It's about `16.66` gas per microsecond of `ecrecover` work. 

With this in mind, `F` took `402` nanoseconds on average, it is equal to `0.402` microseconds so the invocation (assuming the same input parameters) should cost `0.402 * 16.66` = `6.7` gas.

Those benchmarks were done with 12 rounds, so it's `0.56` per round.

In the original EIP there was mentioned a price of `37` gas per round. 

We have it either heavily optimized or I am too tired to see where is the mistake in my calculations...